drivers/usb_c/tcpc/ucpd_numaker.c

```diff
@@ -0,0 +1,2528 @@
+ * @retval -EIO on failure
+ */
+static int numaker_utcpd_bist_test_mode_set_enable(const struct device *dev, bool enable)
+{
+	const struct numaker_tcpc_config *const config = dev->config;
+	UTCPD_T *utcpd_base = config->utcpd_base;
+	int rc;
+	uint32_t ctl = NUMAKER_UTCPD_REG_READ_BY_NAME(dev, CTL);
+
+	/* Enable or not BIST test mode */
+	if (enable) {
+		ctl |= UTCPD_CTL_BISTEN_Msk;
+	} else {
+		ctl &= ~UTCPD_CTL_BISTEN_Msk;
+	}
+	rc = NUMAKER_UTCPD_REG_WRITE_BY_NAME(dev, CTL, ctl);
+	if (rc < 0) {
+		return rc;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Check if UTCPD BIST test mode is enabled
+ *
+ * @retval 1 if UTCPD BIST test mode is enabled
+ * @retval 0 if UTCPD BIST test mode is not enabled
+ */
+static int numaker_utcpd_bist_test_mode_is_enabled(const struct device *dev)
+{
+	const struct numaker_tcpc_config *const config = dev->config;
+	UTCPD_T *utcpd_base = config->utcpd_base;
+	uint32_t ctl = NUMAKER_UTCPD_REG_READ_BY_NAME(dev, CTL);
+
+	if (ctl & UTCPD_CTL_BISTEN_Msk) {
+		return 1;
+	} else {
+		return 0;
+	}
+}
+
+/**
+ * @brief Clears UTCPD Rx message FIFO
+ *
+ * @retval 0 on success
+ */
+static int numaker_utcpd_rx_fifo_clear(const struct device *dev)
+{
+	struct numaker_tcpc_data *data = dev->data;
+
+	data->rx_msg_ready = false;
+
+	return 0;
+}
+
+/**
+ * @brief Enqueues UTCPD Rx message
+ *
+ * @retval 0 on success
+ * @retval -EIO on failure
+ */
+static int numaker_utcpd_rx_fifo_enqueue(const struct device *dev)
+{
+	const struct numaker_tcpc_config *const config = dev->config;
+	struct numaker_tcpc_data *data = dev->data;
+	UTCPD_T *utcpd_base = config->utcpd_base;
+	int rc = 0;
+	uint32_t rxbcnt = NUMAKER_UTCPD_REG_READ_BY_NAME(dev, RXBCNT);
+	uint32_t rxftype = NUMAKER_UTCPD_REG_READ_BY_NAME(dev, RXFTYPE);
+	uint32_t rxhead = NUMAKER_UTCPD_REG_READ_BY_NAME(dev, RXHEAD);
+	uint32_t is = NUMAKER_UTCPD_REG_READ_BY_NAME(dev, IS);
+	uint32_t rx_data_size;
+	struct pd_msg *msg = &data->rx_msg;
+
+	/* Rx message pending? */
+	if (!(is & UTCPD_IS_RXSOPIS_Msk)) {
+		goto cleanup;
+	}
+
+	/* rxbcnt = 1 (frame type) + 2 (Message Header) + Rx data byte count */
+	if (rxbcnt < 3) {
+		LOG_ERR("Invalid UTCPD_RXBCNT: %d", rxbcnt);
+		rc = -EIO;
+		goto cleanup;
+	}
+	rx_data_size = rxbcnt - 3;
+
+	/* Not support Unchunked Extended Message exceeding PD_CONVERT_PD_HEADER_COUNT_TO_BYTES */
+	if (rx_data_size > (PD_MAX_EXTENDED_MSG_LEGACY_LEN + 2)) {
+		LOG_ERR("Not support Unchunked Extended Message exceeding "
+			"PD_CONVERT_PD_HEADER_COUNT_TO_BYTES: %d",
+			rx_data_size);
+		rc = -EIO;
+		goto cleanup;
+	}
+
+	/* Rx FIFO has room? */
+	if (data->rx_msg_ready) {
+		LOG_WRN("Rx FIFO overflow");
+	}
+
+	/* Rx frame type */
+	/* NOTE: Needn't extra cast for UTCPD_RXFTYPE.RXFTYPE aligning with pd_packet_type */
+	msg->type = (rxftype & UTCPD_RXFTYPE_RXFTYPE_Msk) >> UTCPD_RXFTYPE_RXFTYPE_Pos;
+
+	/* Rx header */
+	msg->header.raw_value = (uint16_t)rxhead;
+
+	/* Rx data size */
+	msg->len = rx_data_size;
+
+	/* Rx data */
+	{
```

Comment:
this is odd, maybe split this block in a different function

Reply:
for both Rx/Tx data, create separate functions to handle them